core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -0,0 +1,141 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.TimeoutException;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+final class TimeoutSubscriber<T> implements Runnable, Subscriber<T>, Subscription {
+
+    private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+    private final Subscriber<? super T> delegate;
+    private final EventExecutor executor;
+    private final StreamTimeoutMode timeoutMode;
+    private final Duration timeoutDuration;
+    private final long timeoutNanos;
+    @Nullable
+    private ScheduledFuture<?> timeoutFuture;
+    @Nullable
+    private Subscription subscription;
+    private long lastEventTimeNanos;
+    private boolean completed;
+
+    TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                      StreamTimeoutMode timeoutMode) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.executor = requireNonNull(executor, "executor");
+        this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        timeoutNanos = timeoutDuration.toNanos();
+        this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+    }
+
+    private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+        return executor.schedule(this, delay, unit);
+    }
+
+    void cancelSchedule() {
+        if (timeoutFuture != null && !timeoutFuture.isCancelled()) {
+            timeoutFuture.cancel(false);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+            final long currentTimeNanos = System.nanoTime();
+            final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+            if (elapsedNanos < timeoutNanos) {
+                final long delayNanos = timeoutNanos - elapsedNanos;
+                timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
+                return;
+            }
+        }
+        completed = true;
+        delegate.onError(new TimeoutException(
+                String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
+        subscription.cancel();
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        subscription = s;
+        lastEventTimeNanos = System.nanoTime();
+        timeoutFuture = scheduleTimeout(timeoutNanos, TimeUnit.NANOSECONDS);
```

Comment:
We probably need to schedule the first timeout when `subscription.request();` is called. Could you check how flux implemented it?

Reply:
Had a discussion about this, and we've decided to keep the current implementation because:
- It simplifies the code
- Subscribers probably don't call `subscription.request()` when they encounter issues, so timing out is necessary.